Initium-Odp/src/com/universeprojects/miniup/server/longoperations/LongOperationTeachIdea.java

```diff
@@ -0,0 +1,118 @@
+import com.universeprojects.miniup.server.services.ODPInventionService;
+
+public class LongOperationTeachIdea extends LongOperation{
+
+	public LongOperationTeachIdea(ODPDBAccess db, Map<String, String[]> requestParameters) throws UserErrorMessage {
+		super(db, requestParameters);
+	}
+
+	@Override
+	int doBegin(Map<String, String> parameters) throws UserErrorMessage, UserRequestIncompleteException {
+		int value = 0;
+		
+		CachedEntity character = db.getCurrentCharacter();
+		ODPInventionService ois = db.getInventionService(character, db.getKnowledgeService(character.getKey()));
+		
+		List<Key> ideaKeys = null;
+		
+		String ideaIds = parameters.get("ideaIds");
+		if(ideaIds != null) {
+			ideaKeys = new ArrayList<Key>();
+			for(String ideaId : ideaIds.split(","))
+				ideaKeys.add(KeyFactory.createKey("ConstrucItemIdea", Long.parseLong(ideaId)));
+		}
+		else throw new UserErrorMessage("No ideas were selected.");
+		
+		List<Key> studentKeys = null;
+		
+		String characterIds = parameters.get("characterIds");
+		if(characterIds != null) {
+			studentKeys = new ArrayList<Key>();
+			for(String charId : characterIds.split(","))
+				studentKeys.add(KeyFactory.createKey("Character", Long.parseLong(charId)));
+		}
+		else throw new UserErrorMessage("No students were selected.");
+		
+		db.pool.loadEntities(ideaKeys);
+				
+		List<CachedEntity> ideaEnts = db.pool.get(ideaKeys);
+		List<Key> ideaDefKeys = new ArrayList<Key>();
+		
+		for(CachedEntity idea : ideaEnts) {
+			db.pool.addToQueue(idea.getProperty("_definitionKey"));
+			ideaDefKeys.add((Key) idea.getProperty("_definitionKey"));
+		}
+		
+		List<CachedEntity> ideaDefs = db.pool.get(ideaDefKeys);
+		
+		for(CachedEntity ideaDef : ideaDefs) {
+			//value += ois.getIdeaTeachingTime(character, null, ideaDef);
+		}
+		
+		setDataProperty("ideaDefKeys", ideaDefKeys);
+		setDataProperty("studentKeys", studentKeys);
+		
+		return value;
+	}
+
+	@Override
+	String doComplete() throws UserErrorMessage, UserRequestIncompleteException, ContinuationException {
+		CachedEntity character = db.getCurrentCharacter();
+		
+		List<Key> studentKeys = (List<Key>) getDataProperty("studentKeys");
+		List<Key> ideaDefKeys = (List<Key>) getDataProperty("ideaDefKeys");
+		
+		db.pool.addToQueue(studentKeys);
+		db.pool.addToQueue(ideaDefKeys);
+		
+		List<CachedEntity> students = db.pool.get(studentKeys);
+		List<CachedEntity> ideaDefs = db.pool.get(ideaDefKeys);
+		
+		QueryHelper query = new QueryHelper(db.getDB());
+		
+		for(CachedEntity ideaDef : ideaDefs) {
+			nextStudent:
+			for(CachedEntity student : students) {
+				//make sure the student is still valid. Ensure that they're still in learn mode, and that they're still in the same location.
+				
+				if(GameUtils.equals(student.getProperty("locationKey"), character.getProperty("locationKey")) == false) continue;
+				if(GameUtils.equals(student.getProperty("mode"), "LEARNING") == false) continue; //not sure if this is how we want to do this.
```

Comment:
I don't think the player enters a different mode when doing an invention do they? So we would want to do it just like invention.

Reply:
Normal inventions are just a long operation, which locks the character inventing. But in this case, we are affecting other characters. How do we mark the characters that are being taught, to prevent them from doing other things while they are being taught?